File: Sources/VoxLocalCore/Refinement/OllamaRefinementProvider.swift
import Foundation

/// Refines transcripts through a local Ollama server. Privacy guarantee:
/// the endpoint must resolve to a loopback host, otherwise every call fails
/// with `RefinementError.nonLocalEndpoint` — no request leaves the machine.
public final class OllamaRefinementProvider: TextRefinementProvider, @unchecked Sendable {
    public let endpoint: URL
    public let model: String
    private let session: URLSession

    static let loopbackHosts: Set<String> = ["127.0.0.1", "localhost", "::1", "[::1]"]

    /// Refuses every HTTP redirect. Without this, a process answering on
    /// the loopback port could 307-redirect the transcript POST to a remote
    /// host and URLSession would follow it — breaking the privacy guarantee.
    private final class NoRedirectDelegate: NSObject, URLSessionTaskDelegate {
        func urlSession(_ session: URLSession, task: URLSessionTask,
                        willPerformHTTPRedirection response: HTTPURLResponse,
                        newRequest request: URLRequest,
                        completionHandler: @escaping @Sendable (URLRequest?) -> Void) {
            completionHandler(nil) // deliver the 3xx as-is; never follow
        }
    }

    public init(endpoint: String, model: String, session: URLSession? = nil) throws {
        guard let url = URL(string: endpoint), let host = url.host else {
            throw RefinementError.nonLocalEndpoint(endpoint)
        }
        guard Self.isLoopback(host: host) else {
            throw RefinementError.nonLocalEndpoint(endpoint)
        }
        self.endpoint = url
        self.model = model
        if let session {
            self.session = session
        } else {
            let config = URLSessionConfiguration.ephemeral
            config.timeoutIntervalForRequest = 60
            config.timeoutIntervalForResource = 120
            self.session = URLSession(configuration: config, delegate: NoRedirectDelegate(), delegateQueue: nil)
        }
    }

    public static func isLoopback(host: String) -> Bool {
        loopbackHosts.contains(host.lowercased())
    }

    // MARK: - Availability

    private struct TagsResponse: Decodable {
        struct Model: Decodable { let name: String }
        let models: [Model]?
    }

    public func checkAvailability() async -> RefinementAvailability {
        do {
            let names = try await installedModels()
            if names.contains(where: { $0 == model || $0.hasPrefix(model + ":") }) {
                return .available
            }
            return .modelMissing(available: names)
        } catch {
            return .serverUnreachable(error.localizedDescription)
        }
    }

    /// Model names installed on the local Ollama server (`/api/tags`).
    public func installedModels() async throws -> [String] {
        var request = URLRequest(url: endpoint.appendingPathComponent("api/tags"))
        request.timeoutInterval = 5
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            throw RefinementError.unavailable("HTTP \((response as? HTTPURLResponse)?.statusCode ?? -1)")
        }
        let decoded = try JSONDecoder().decode(TagsResponse.self, from: data)
        return (decoded.models ?? []).map(\.name)
    }

    // MARK: - Refinement

    struct ChatRequest: Encodable {
        struct Message: Encodable {
            let role: String
            let content: String
        }
        struct Options: Encodable {
            let temperature: Double
            let num_predict: Int
        }
        let model: String
        let messages: [Message]
        let stream: Bool
        let options: Options
    }

    private struct ChatResponse: Decodable {
        struct Message: Decodable { let content: String? }
        let message: Message?
    }

    /// Builds the `/api/chat` request body (exposed for unit tests).
    public static func requestBody(model: String, transcript: String, context: RefinementContext) throws -> Data {
        let request = ChatRequest(
            model: model,
            messages: [
                .init(role: "system", content: RefinementPromptBuilder.systemPrompt(for: context)),
                .init(role: "user", content: RefinementPromptBuilder.userPrompt(transcript: transcript)),
            ],
            stream: false,
            // Cap generation relative to input so a runaway model cannot
            // produce unbounded output.
            options: .init(temperature: 0.2, num_predict: max(256, transcript.count)))
        return try JSONEncoder().encode(request)
    }

    public func refine(_ transcript: String, context: RefinementContext) async throws -> String {
        var request = URLRequest(url: endpoint.appendingPathComponent("api/chat"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try Self.requestBody(model: model, transcript: transcript, context: context)
        request.timeoutInterval = context.timeout

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch let error as URLError where error.code == .timedOut {
            throw RefinementError.timeout
        } catch let error as URLError where error.code == .cancelled {
            throw CancellationError()
        } catch {
            throw RefinementError.unavailable(error.localizedDescription)
        }

        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw RefinementError.invalidResponse("HTTP \(code)")
        }
        guard let decoded = try? JSONDecoder().decode(ChatResponse.self, from: data),
              let content = decoded.message?.content else {
            throw RefinementError.invalidResponse("unexpected /api/chat payload")
        }
        return content
    }
}

File: windows/src/VoxLocal.App/Core/Refinement/OllamaRefinementProvider.cs
using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace VoxLocal.Core.Refinement;

/// <summary>
/// Refines transcripts through a local Ollama server. Privacy guarantee:
/// the endpoint must resolve to a loopback host, otherwise construction
/// fails with NonLocalEndpointException — no request leaves the machine.
/// </summary>
public sealed class OllamaRefinementProvider : ITextRefinementProvider
{
    public Uri Endpoint { get; }
    public string Model { get; }
    private readonly HttpClient _http;

    internal static readonly HashSet<string> LoopbackHosts = new(StringComparer.OrdinalIgnoreCase)
    {
        "127.0.0.1", "localhost", "::1", "[::1]",
    };

    public OllamaRefinementProvider(string endpoint, string model, HttpClient? http = null)
    {
        if (!Uri.TryCreate(endpoint, UriKind.Absolute, out var url) || string.IsNullOrEmpty(url.Host))
        {
            throw new NonLocalEndpointException(endpoint);
        }
        if (!IsLoopback(url.Host))
        {
            throw new NonLocalEndpointException(endpoint);
        }
        Endpoint = url;
        Model = model;
        // Refuse every HTTP redirect. Without this, a process answering on
        // the loopback port could 307-redirect the transcript POST to a
        // remote host — AllowAutoRedirect=false delivers the 3xx as-is and
        // never follows it (same guarantee as NoRedirectDelegate on macOS).
        _http = http ?? new HttpClient(new SocketsHttpHandler
        {
            AllowAutoRedirect = false,
            UseCookies = false,
        })
        {
            Timeout = TimeSpan.FromSeconds(120),
        };
    }

    public static bool IsLoopback(string host) => LoopbackHosts.Contains(host);

    // ---- Availability ----

    private sealed class TagsResponse
    {
        [JsonPropertyName("models")]
        public List<ModelEntry>? Models { get; set; }

        public sealed class ModelEntry
        {
            [JsonPropertyName("name")]
            public string Name { get; set; } = "";
        }
    }

    public async Task<RefinementAvailability> CheckAvailabilityAsync()
    {
        try
        {
            var names = await InstalledModelsAsync().ConfigureAwait(false);
            if (names.Any(n => n == Model || n.StartsWith(Model + ":", StringComparison.Ordinal)))
            {
                return new RefinementAvailability.Available();
            }
            return new RefinementAvailability.ModelMissing(names);
        }
        catch (Exception e)
        {
            return new RefinementAvailability.ServerUnreachable(e.Message);
        }
    }

    /// <summary>Model names installed on the local Ollama server (/api/tags).</summary>
    public async Task<IReadOnlyList<string>> InstalledModelsAsync(CancellationToken cancellationToken = default)
    {
        using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        cts.CancelAfter(TimeSpan.FromSeconds(5));
        using var response = await _http.GetAsync(new Uri(Endpoint, "api/tags"), cts.Token).ConfigureAwait(false);
        if ((int)response.StatusCode != 200)
        {
            throw new RefinementUnavailableException($"HTTP {(int)response.StatusCode}");
        }
        var json = await response.Content.ReadAsStringAsync(cts.Token).ConfigureAwait(false);
        var decoded = JsonSerializer.Deserialize<TagsResponse>(json);
        return (decoded?.Models ?? new()).Select(m => m.Name).ToList();
    }

    // ---- Refinement ----

    internal sealed record ChatMessage(
        [property: JsonPropertyName("role")] string Role,
        [property: JsonPropertyName("content")] string Content);

    internal sealed record ChatOptions(
        [property: JsonPropertyName("temperature")] double Temperature,
        [property: JsonPropertyName("num_predict")] int NumPredict);

    internal sealed record ChatRequest(
        [property: JsonPropertyName("model")] string Model,
        [property: JsonPropertyName("messages")] IReadOnlyList<ChatMessage> Messages,
        [property: JsonPropertyName("stream")] bool Stream,
        [property: JsonPropertyName("options")] ChatOptions Options);

    private sealed class ChatResponse
    {
        [JsonPropertyName("message")]
        public ResponseMessage? Message { get; set; }

        public sealed class ResponseMessage
        {
            [JsonPropertyName("content")]
            public string? Content { get; set; }
        }
    }

    /// <summary>Builds the /api/chat request body (exposed for unit tests).</summary>
    public static byte[] RequestBody(string model, string transcript, RefinementContext context)
    {
        var request = new ChatRequest(
            model,
            new[]
            {
                new ChatMessage("system", RefinementPromptBuilder.SystemPrompt(context)),
                new ChatMessage("user", RefinementPromptBuilder.UserPrompt(transcript)),
            },
            Stream: false,
            // Cap generation relative to input so a runaway model cannot
            // produce unbounded output.
            new ChatOptions(Temperature: 0.2, NumPredict: Math.Max(256, transcript.Length)));
        return JsonSerializer.SerializeToUtf8Bytes(request);
    }

    public async Task<string> RefineAsync(
        string transcript,
        RefinementContext context,
        CancellationToken cancellationToken = default)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, new Uri(Endpoint, "api/chat"))
        {
            Content = new ByteArrayContent(RequestBody(Model, transcript, context)),
        };
        request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/json");

        using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        cts.CancelAfter(TimeSpan.FromSeconds(context.TimeoutSeconds));

        HttpResponseMessage response;
        string body;
        try
        {
            response = await _http.SendAsync(request, cts.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            throw; // session cancelled — the pipeline maps this to a fallback
        }
        catch (OperationCanceledException)
        {
            throw new RefinementTimeoutException();
        }
        catch (HttpRequestException e)
        {
            throw new RefinementUnavailableException(e.Message);
        }

        using (response)
        {
            if ((int)response.StatusCode != 200)
            {
                throw new RefinementInvalidResponseException($"HTTP {(int)response.StatusCode}");
            }
            body = await response.Content.ReadAsStringAsync(cts.Token).ConfigureAwait(false);
        }

        ChatResponse? decoded;
        try
        {
            decoded = JsonSerializer.Deserialize<ChatResponse>(body);
        }
        catch (JsonException)
        {
            decoded = null;
        }
        if (decoded?.Message?.Content is not { } content)
        {
            throw new RefinementInvalidResponseException("unexpected /api/chat payload");
        }
        return content;
    }
}
